Failure to parse media type

Caused by: java.lang.IllegalArgumentException: Could not parse '*; q=.2'
	at io.helidon.common.http.MediaType.parse(MediaType.java:253)

Request from another service using Jersey client with default Accept header.

--- a/webserver/webserver/src/main/java/io/helidon/webserver/HashRequestHeaders.java
+++ b/webserver/webserver/src/main/java/io/helidon/webserver/HashRequestHeaders.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017, 2018 Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2017, 2018, 2019 Oracle and/or its affiliates. All rights reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -110,8 +110,8 @@ class HashRequestHeaders extends ReadOnlyParameters implements RequestHeaders {
         List<MediaType> result = this.acceptedtypesCache;
         if (result == null) {
             List<String> acceptValues = all(Http.Header.ACCEPT);
-            result = acceptValues.size() == 1 && HUC_ACCEPT_DEFAULT.equals(acceptValues.get(0)) ?
-                    HUC_ACCEPT_DEFAULT_TYPES : acceptValues.stream()
+            result = acceptValues.size() == 1 && HUC_ACCEPT_DEFAULT.equals(acceptValues.get(0))
+                    ? HUC_ACCEPT_DEFAULT_TYPES : acceptValues.stream()
                             .flatMap(h -> Utils.tokenize(',', "\"", false, h).stream())
                             .map(String::trim)
                             .map(MediaType::parse)

--- a/webserver/webserver/src/main/java/io/helidon/webserver/HashRequestHeaders.java
+++ b/webserver/webserver/src/main/java/io/helidon/webserver/HashRequestHeaders.java
@@ -26,6 +26,7 @@ import java.util.Optional;
 import java.util.OptionalLong;
 import java.util.stream.Collectors;
 
+import io.helidon.common.CollectionsHelper;
 import io.helidon.common.http.Http;
 import io.helidon.common.http.MediaType;
 import io.helidon.common.http.Parameters;
@@ -36,6 +37,22 @@ import io.helidon.common.http.Utils;
  * A {@link RequestHeaders} implementation on top of {@link ReadOnlyParameters}.
  */
 class HashRequestHeaders extends ReadOnlyParameters implements RequestHeaders {
+
+    /**
+     * Header value of the non compliant {@code Accept} header sent by
+     * {@link HTTPURLConnection} when none is set.
+     * @see <a href="https://bugs.openjdk.java.net/browse/JDK-8163921">JDK-8163921</a>
+     */
+    static final String HUC_ACCEPT_DEFAULT = "text/html, image/gif, image/jpeg, *; q=.2, */*; q=.2";
+
+    /**
+     * Accepted types for {@link #HUC_ACCEPT_DEFAULT}.
+     */
+    private static final List<MediaType> HUC_ACCEPT_DEFAULT_TYPES = CollectionsHelper.listOf(
+                MediaType.TEXT_HTML,
+                MediaType.parse("image/gif"),
+                MediaType.parse("image/jpeg"),
+                MediaType.parse("*/*; q=.2"));
 
     private final Object internalLock = new Object();
     private volatile Parameters cookies;
@@ -92,7 +109,9 @@ class HashRequestHeaders extends ReadOnlyParameters implements RequestHeaders {
     public List<MediaType> acceptedTypes() {
         List<MediaType> result = this.acceptedtypesCache;
         if (result == null) {
-            result = all(Http.Header.ACCEPT).stream()
+            List<String> acceptValues = all(Http.Header.ACCEPT);
+            result = acceptValues.size() == 1 && HUC_ACCEPT_DEFAULT.equals(acceptValues.get(0)) ?
+                    HUC_ACCEPT_DEFAULT_TYPES : acceptValues.stream()
                             .flatMap(h -> Utils.tokenize(',', "\"", false, h).stream())
                             .map(String::trim)
                             .map(MediaType::parse)

--- a/webserver/webserver/src/test/java/io/helidon/webserver/HashRequestHeadersTest.java
+++ b/webserver/webserver/src/test/java/io/helidon/webserver/HashRequestHeadersTest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017, 2018 Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2017, 2018, 2019 Oracle and/or its affiliates. All rights reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.

--- a/webserver/webserver/src/test/java/io/helidon/webserver/HashRequestHeadersTest.java
+++ b/webserver/webserver/src/test/java/io/helidon/webserver/HashRequestHeadersTest.java
@@ -37,6 +37,7 @@ import static org.hamcrest.CoreMatchers.is;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.collection.IsIterableContainingInOrder.contains;
 import static org.hamcrest.core.IsCollectionContaining.hasItems;
+import org.junit.jupiter.api.Assertions;
 
 /**
  * Tests {@link HashRequestHeaders}.
@@ -102,6 +103,26 @@ public class HashRequestHeadersTest {
         assertThat(hs.acceptedTypes().get(2), is(createMt("text", "html", CollectionsHelper.mapOf("level", "1"))));
         assertThat(hs.acceptedTypes().get(3),
                    is(createMt("text", "html", CollectionsHelper.mapOf("level", "2", "q", "0.4"))));
+    }
+
+    @Test
+    public void hucDefaultAccept(){
+        try {
+            HashRequestHeaders hs = withHeader(Http.Header.ACCEPT, HashRequestHeaders.HUC_ACCEPT_DEFAULT);
+            assertThat(hs.acceptedTypes().get(0), is(MediaType.TEXT_HTML));
+            assertThat(hs.acceptedTypes().get(1), is(createMt("image", "gif")));
+            assertThat(hs.acceptedTypes().get(2), is(createMt("image", "jpeg")));
+            assertThat(hs.acceptedTypes().get(3), is(createMt("*", "*", CollectionsHelper.mapOf("q", ".2"))));
+        } catch(IllegalStateException ex){
+            Assertions.fail(ex.getMessage(), ex);
+        }
+    }
+
+    private MediaType createMt(String type, String subtype) {
+        return MediaType.builder()
+                .type(type)
+                .subtype(subtype)
+                .build();
     }
 
     private MediaType createMt(String type, String subtype, Map<String, String> params) {

